File: src/main/java/com/niteshsynergy/dsa/Week16/PairSumSortedNoSortedArray.java
package com.niteshsynergy.dsa.Week16;

import java.util.*;
import java.util.stream.IntStream;

/*
Developed By Nitesh Synergy on 27-08-2025
 */
public class PairSumSortedNoSortedArray {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter size of array: ");
        int n = sc.nextInt();
        int[] arr = new int[n];
        System.out.println("Enter array: ");
        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();
        System.out.println("Enter Target Element: ");
        int target = sc.nextInt();

    }
    public static boolean pairSumSortedArray(int[] arr, int target) {
        // assume here Array is sorted.

        int left = 0, right = arr.length - 1;

        while (left < right) {
            int sum = arr[left] + arr[right];
            if(sum==target)
                return true;
            else if(sum>target)
                right--;
            else
                left++;
        }
        return false;
    }
    public static boolean hasPairWithSumStreamApi(int[] arr, int target) {
        int n = arr.length;

        return IntStream.range(0, n)  // i from 0 to n-1
                .anyMatch(i ->
                        IntStream.range(i + 1, n)  // j from i+1 to n-1
                                .anyMatch(j -> arr[i] + arr[j] == target)
                );
    }
    public static boolean hasPairWithSumStreamApiWithSet(int[] arr, int target) {
        Set<Integer> seen = new HashSet<>();

        return Arrays.stream(arr)
                .anyMatch(num -> {
                    if (seen.contains(target - num)) {
                        return true;
                    }
                    seen.add(num);
                    return false;
                });
    }
    public static boolean hasPairWithSumUsingStreamWithBoxedFlatMap(int[] arr, int target) {
        int n = arr.length;

        return IntStream.range(0, n)
                .boxed()
                .flatMap(i -> IntStream.range(i + 1, n)
                        .mapToObj(j -> arr[i] + arr[j]))
                .anyMatch(sum -> sum == target);
    }
    public static boolean hasPairWithSumUsingStreamHashMap(int[] arr, int target) {
        Map<Integer, Integer> freq = new HashMap<>();

        return Arrays.stream(arr).anyMatch(num -> {
            int complement = target - num;
            if (freq.getOrDefault(complement, 0) > 0) {
                return true;
            }
            freq.put(num, freq.getOrDefault(num, 0) + 1);
            return false;
        });
    }
    // HashSet Approach (O(n), works for unsorted arrays too)
    public static boolean hasPairWithSumUsingHashSet(int[] arr, int target) {
        Set<Integer> seen = new HashSet<>();

        for (int num : arr) {
            int complement = target - num;
            if (seen.contains(complement)) {
                return true; // found pair
            }
            seen.add(num);
        }
        return false;
    }
    //HashMap Approach (O(n), handles duplicates well)
    public static boolean hasPairWithSumUseHashMapWithDuplicate(int[] arr, int target) {
        Map<Integer, Integer> freq = new HashMap<>();

        for (int num : arr) {
            int complement = target - num;
            if (freq.getOrDefault(complement, 0) > 0) {
                return true; // found complement
            }
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }
        return false;
    }
//Brute Force Approach (O(n²))
    public static boolean hasPairWithSumBruteForce(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[i] + arr[j] == target) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: src/main/java/com/niteshsynergy/dsa/Week16/PairSumSortedNoSortedArray.py
from typing import List

# 1. Two-pointer (for sorted arrays)
def pair_sum_sorted_array(arr: List[int], target: int) -> bool:
    left, right = 0, len(arr) - 1
    while left < right:
        s = arr[left] + arr[right]
        if s == target:
            return True
        elif s > target:
            right -= 1
        else:
            left += 1
    return False


# 2. Brute Force (O(n²))
def has_pair_with_sum_bruteforce(arr: List[int], target: int) -> bool:
    n = len(arr)
    for i in range(n):
        for j in range(i+1, n):
            if arr[i] + arr[j] == target:
                return True
    return False


# 3. HashSet Approach (O(n))
def has_pair_with_sum_hashset(arr: List[int], target: int) -> bool:
    seen = set()
    for num in arr:
        complement = target - num
        if complement in seen:
            return True
        seen.add(num)
    return False


# 4. HashMap Approach (handles duplicates) (O(n))
def has_pair_with_sum_hashmap(arr: List[int], target: int) -> bool:
    freq = {}
    for num in arr:
        complement = target - num
        if freq.get(complement, 0) > 0:
            return True
        freq[num] = freq.get(num, 0) + 1
    return False


# 5. Using Python any() with generator expression (stream-like)
def has_pair_with_sum_generator(arr: List[int], target: int) -> bool:
    return any(arr[i] + arr[j] == target for i in range(len(arr)) for j in range(i+1, len(arr)))


# 6. Functional style using itertools.combinations (brute force but clean)
from itertools import combinations
def has_pair_with_sum_itertools(arr: List[int], target: int) -> bool:
    return any(x + y == target for x, y in combinations(arr, 2))
